gh-129210: GC optimisations for case where no objects being collected have finalizers

duaneg opened the pull request:
If no objects being collected have finalizers, they will not be able to be resurrected. This lets us avoid some repeated work we would otherwise have to do.

<!-- gh-issue-number: gh-129210 -->
* Issue: gh-129210
<!-- /gh-issue-number -->

duaneg committed:
gh-129210: Skip second reachability check if no finalizers run

To handle objects potentially resurrected in finalizers it is necessary to
repeat the reachability check on objects to be collected. If there are no
finalizers run this can be skipped: in that case it is not possible for an
object to be resurrected.

In my environment and testing this gives a ~25-30% performance improvement for
the gh-129210 micro-benchmark.

duaneg committed:
gh-129210: Avoid unnecessary list init & merge

Avoid an unnecessary list initialisation and merge: this is a very marginal
optimisation which gains a couple of percent speed increase for a few
benchmarks.

The marginal performance improvement may not be worth the marginal additional
code complexity.

ZeroIntensity reviewed (commented):
Are there any benchmarks? (Also, please don't force push and make the bot spam; we squash merge at the end, and you need a blurb entry anyway.)

duaneg committed:
Add blurb

duaneg committed:
gh-129210: Skip second reachability check in ft GC if no finalizers run

As for the regular GIL garbage collector, skip the second reachability check if
no objects to-be-collected have finalizers, and hence none can be resurrected.

In (so far very limited) testing this gives a ~10% overall performance
improvement to the motivating benchmark's runtime and a ~35% improvement to its
reported time.

The pyperformance gc_traversal and gc_collect benchmarks show no significant
difference and a 6% improvement respectively.

JelleZijlstra commented on Misc/NEWS.d/next/Core_and_Builtins/2025-04-16-12-02-26.gh-issue-129210.a5uLpE.rst:
```diff
@@ -0,0 +1,2 @@
+Optimise garbage collection in case where objects to collect have
```
```suggestion
Optimise garbage collection in case where no objects to collect have
```

JelleZijlstra commented on Python/gc.c:
```diff
@@ -1074,11 +1074,12 @@ handle_legacy_finalizers(PyThreadState *tstate,
  * Note that this may remove some (or even all) of the objects from the
  * list, due to refcounts falling to 0.
```
```suggestion
 * list, due to refcounts falling to 0.
 * Return 1 if any finalizers were run.
```

JelleZijlstra commented on Python/gc_free_threading.c:
```diff
@@ -1593,22 +1593,25 @@ debug_cycle(const char *msg, PyObject *op)
  * Note that this may remove some (or even all) of the objects from the
  * list, due to refcounts falling to 0.
```
```suggestion
 * list, due to refcounts falling to 0.
 * Return 1 if any finalizers were run.
```

JelleZijlstra reviewed (commented):
Makes sense but I'm not familiar enough with the GC to approve.

duaneg committed:
Update Misc/NEWS.d/next/Core_and_Builtins/2025-04-16-12-02-26.gh-issue-129210.a5uLpE.rst

Co-authored-by: Jelle Zijlstra <jelle.zijlstra@gmail.com>

duaneg committed:
Update Python/gc.c

Co-authored-by: Jelle Zijlstra <jelle.zijlstra@gmail.com>

duaneg committed:
Update Python/gc_free_threading.c

Co-authored-by: Jelle Zijlstra <jelle.zijlstra@gmail.com>
